qiskit/primitives/estimator.py

```diff
@@ -126,3 +130,35 @@ def _call(
 
     def close(self):
         self._is_closed = True
+
+    def _run(
+        self,
+        circuits: Sequence[QuantumCircuit],
+        observables: Sequence[BaseOperator | PauliSumOp],
+        parameter_values: Sequence[Sequence[float]],
+        **run_options,
+    ) -> PrimitiveJob:
+        circuit_indices = []
+        for circuit in circuits:
+            index = self._circuit_ids.get(id(circuit))
+            if index is not None:
+                circuit_indices.append(index)
+            else:
+                circuit_indices.append(len(self._circuits))
+                self._circuit_ids[id(circuit)] = len(self._circuits)
+                self._circuits.append(circuit)
```

Comment:
Personally I'd use an `OrderedDict` that has the circuit ids as the keys and circuits as the values. This allowed a single variable instead of having to have both `_circuits` and `_circuit_ids`. An`OrderedDict` also allows accessing by index (e.g. `od.values()[0]` to support the deprecated circuit indexes. 

The down side is you'd probably have to refactor the `_call` method.

Reply:
Now dict is ordered, so we can simply use dict. But it takes O(n) to find index.
In the future (after deprecation of `__call__`), we don't need ids, so I want to refactor after that.